protocol/kafka_confluent/v2/option.go

```diff
@@ -0,0 +1,158 @@
+/*
+ Copyright 2023 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package kafka_confluent
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
+)
+
+// Option is the function signature required to be considered an kafka_confluent.Option.
+type Option func(*Protocol) error
+
+func WithConfigMap(config *kafka.ConfigMap) Option {
+	return func(p *Protocol) error {
+		if config == nil {
+			return fmt.Errorf("the kafka.ConfigMap option must not be nil")
+		}
+		p.kafkaConfigMap = config
+		return nil
+	}
+}
+
+// WithSenderTopic sets the defaultTopic for the kafka.Producer. This option is not required.
+func WithSenderTopic(defaultTopic string) Option {
+	return func(p *Protocol) error {
+		if defaultTopic == "" {
+			return fmt.Errorf("the producer topic option must not be nil")
+		}
+		p.producerDefaultTopic = defaultTopic
+		return nil
+	}
+}
+
+// WithDeliveryChan sets the deliveryChan for the kafka.Producer. This option is not required.
+func WithDeliveryChan(deliveryChan chan kafka.Event) Option {
+	return func(p *Protocol) error {
+		if deliveryChan == nil {
+			return fmt.Errorf("the producer deliveryChan option must not be nil")
+		}
+		p.producerDeliveryChan = deliveryChan
+		return nil
+	}
+}
+
+func WithReceiverTopics(topics []string) Option {
+	return func(p *Protocol) error {
+		if topics == nil {
+			return fmt.Errorf("the consumer topics option must not be nil")
+		}
+		p.consumerTopics = topics
+		return nil
+	}
+}
+
+func WithRebalanceCallBack(rebalanceCb kafka.RebalanceCb) Option {
+	return func(p *Protocol) error {
+		if rebalanceCb == nil {
+			return fmt.Errorf("the consumer group rebalance callback must not be nil")
+		}
+		p.consumerRebalanceCb = rebalanceCb
+		return nil
+	}
+}
+
+func WithPollTimeout(timeoutMs int) Option {
+	return func(p *Protocol) error {
+		p.consumerPollTimeout = timeoutMs
+		return nil
+	}
+}
+
+func WithSender(producer *kafka.Producer) Option {
+	return func(p *Protocol) error {
+		if producer == nil {
+			return fmt.Errorf("the producer option must not be nil")
+		}
+		p.producer = producer
+		return nil
+	}
+}
+
+func WithReceiver(consumer *kafka.Consumer) Option {
+	return func(p *Protocol) error {
+		if consumer == nil {
+			return fmt.Errorf("the consumer option must not be nil")
+		}
+		p.consumer = consumer
+		return nil
+	}
+}
+
+// Opaque key type used to store offsets: assgin offset from ctx, commit offset from context
+type commitOffsetType struct{}
+
+var offsetKey = commitOffsetType{}
+
+// WithOffsetCtx will return the topic partitions to commit offsets for.
```

Comment:
nit: this is a modify operation, so the comment seems a bit off

Reply:
also, I found the name to be a bit misleading:

1) remove the `Ctx` suffix to follow option style
2) There's no offset parameter in the signature (hidden deep inside topicPartitions)

IIUC what this does it sets the offsets to start consuming from, per topic/partition. If so, rename parameter to `topicPartitionOffsets`. Also, some validations might make sense e.g., checking if all three properties are set in each slice value.

Reply:
Done